### PR TITLE
EGL Rework

### PR DESCRIPTION
**Pull request by Drakulix**
Alright, it is late and I am not sure, if I did not miss anything important.
Docs are also largely missing.
Other then that, I think this is good for review/first feedback.

Partly addresses #12.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -5,22 +5,27 @@ authors = ["Victor Berger <victor.berger@thalesgroup.com>"]
...
-wayland-server = "0.9.1"
+wayland-server = "0.9.4"
 nix = "0.7.0"
 xkbcommon = "0.2.1"
 tempfile = "2.1.5"
 slog = { version = "2.0.0" }
 slog-stdlog = "2.0.0-0.2"
-glutin = { version = "~0.7.4", optional = true }
+libloading = "0.4.0"
+wayland-client = { version = "~0.8.6", optional = true }
```
Comment (vberger) [inline Cargo.toml]: Having wayland-client 0.8 and wayland-server 0.9 actually compiles? O.o

File: Cargo.toml [hunk 2]
```diff
@@ -5,22 +5,27 @@ authors = ["Victor Berger <victor.berger@thalesgroup.com>"]
...
+gl_generator = "0.5"
+
 [dev-dependencies]
 slog-term = "~1.5"
 
 [features]
-default = ["backend_glutin", "backend_libinput", "renderer_glium"]
-backend_glutin = ["glutin", "wayland-server/dlopen"]
+default = ["backend_winit", "backend_libinput", "renderer_glium"]
+backend_winit = ["winit", "wayland-server/dlopen", "wayland-client", "wayland-client/dlopen"]
```
Comment (vberger) [inline Cargo.toml]: I don't think we need both `wayland-client` and `wayland-client/dlopen` here, do we?

File: src/backend/glutin.rs [hunk 3]
```diff
@@ -21,7 +21,7 @@ use std::rc::Rc;
 /// Create a new `GlutinHeadlessRenderer` which implements the `OpenglRenderer` graphics
 /// backend trait
 pub fn init_headless_renderer() -> Result<GlutinHeadlessRenderer, CreationError> {
-    init_headless_renderer_from_builder(HeadlessRendererBuilder::new(1024, 600))
+    init_headless_renderer_from_builder(HeadlessRendererBuilder::new(1024, 600)
```
Comment (vberger) [inline src/backend/glutin.rs]: unclosed delimiter `(` 😛

File: src/backend/graphics/egl.rs [hunk 4]
```diff
@@ -0,0 +1,662 @@
+//! Common traits and types for opengl rendering on graphics backends
+
+/// Large parts of the following file are taken from
+/// https://github.com/tomaka/glutin/tree/master/src/api/egl at commit
+/// `044e651edf67a2029eecc650dd42546af1501414`
+///
+/// It therefor falls under glutin's Apache 2.0 license
+/// (see https://github.com/tomaka/glutin/blob/master/LICENSE)
```
Comment (vberger) [inline src/backend/graphics/egl.rs]: "therefor**e**"

Also I guess this should be good enough. But to be safe, let's ask @tomaka his opinion on the matter.

File: src/backend/graphics/egl.rs [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+
+        include!(concat!(env!("OUT_DIR"), "/egl_bindings.rs"));
+    }
+}
+
+#[derive(Clone, Copy)]
+pub enum Native {
+    X11(ffi::NativeDisplayType, ffi::NativeWindowType),
+    Wayland(ffi::NativeDisplayType, ffi::NativeWindowType),
+    Gbm(ffi::NativeDisplayType, ffi::NativeWindowType),
```
Comment (vberger) [inline src/backend/graphics/egl.rs]: Is this what will be used for the native/modesetting backed?

File: src/backend/winit.rs [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+    }
+
+    /// Processes new events of the underlying event loop to drive the set `InputHandler`.
+    ///
+    /// You need to periodically call this function to keep the underlying event loop and
+    /// `Window` active. Otherwise the window may no respond to user interaction and no
+    /// input events will be received by a set `InputHandler`.
+    ///
+    /// Returns an error if the `Window` the window has been closed. Calling
+    /// `process_new_events` again after the `Window` has been closed is considered an
```
Comment (vberger) [inline src/backend/winit.rs]: `dispatch_new_events`

File: src/backend/glutin.rs (hunk 3 above)
Comment (vberger) [inline src/backend/glutin.rs]: Actually, is this file needed at all now?

**Comment by vberger** [comment]
I don't have much to say other than this. This is quite a lot of boilerplate to plug everything together, not the easiest thing to review. I'll try to give a few more passes on it.

Also, making the travis build pass would be a good thing to do as well. 😉

File: src/backend/graphics/egl.rs (hunk 4 above)
Comment (tomaka) [inline src/backend/graphics/egl.rs]: Well, yeah, I don't see a problem with this.

File: Cargo.toml (hunk 1 above)
Comment (Drakulix) [inline Cargo.toml]: Yes when I set both to use dlopen.
And wayland-client 0.8 is sadly required, as winit is still expecting that.

File: Cargo.toml (hunk 2 above)
Comment (Drakulix) [inline Cargo.toml]: See comment above.

File: src/backend/glutin.rs (hunk 3 above)
Comment (Drakulix) [inline src/backend/glutin.rs]: No, I just forgot to remove it.

File: src/backend/graphics/egl.rs (hunk 5 above)
Comment (Drakulix) [inline src/backend/graphics/egl.rs]: Yes. This was the main motivation of doing this change now, as this is essential to make anything useful out of the drm implementation I am working on.

File: Cargo.toml (hunk 2 above)
Comment (vberger) [inline Cargo.toml]: But, don't `wayland-client/dlopen` imply `wayland-client` ?

**Comment by Drakulix** [comment]
Thanks for the first review. Typos, Docs and Travis are definitely my main priority today.

File: Cargo.toml (hunk 2 above)
Comment (Drakulix) [inline Cargo.toml]: Not sure, good point. I will test that.

**Comment by Drakulix** [comment]
Alright, I should have addressed most of the things.

The example is still not working right (immediately closes on wayland and panics on x11), but its getting somewhere.

**Comment by vberger** [comment]
> The example is still not working right (immediately closes on wayland and panics on x11), but its getting somewhere.

Closes without panicking on wayland? Well, that's weird, is it some segfault or stuff like that?

File: src/backend/winit.rs [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+    let events_loop = EventsLoop::new();
+    let window = Rc::new(builder.build(&events_loop)?);
+
+    let (native, surface) = if let (Some(conn), Some(window)) =
+        (get_x11_xconnection(), window.get_xlib_window()) {
+        (Native::X11(conn.display as *const _, window), None)
+    } else if let (Some(display), Some(surface)) =
+        (window.get_wayland_display(), window.get_wayland_client_surface()) {
+        let (w, h) = window.get_inner_size().unwrap();
+        let egl_surface = wegl::WlEglSurface::new(surface, w as i32, h as i32);
```
Comment (vberger) [inline src/backend/winit.rs]: You'll want to keep this `egl_surface` around, as dropping it destroys the EGL_surface, and you'll need it to handle resizing of the surface.

File: src/backend/winit.rs [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+            let surface = &self.surface;
+            let mut handler = self.handler.as_mut();
+
+            self.events_loop
+                .poll_events(move |event| if let Some(ref mut handler) = handler {
+                                 let Event::WindowEvent { event, .. } = event;
+                                 match event {
+                                     WindowEvent::Resized(x, y) => {
+                                         window.set_inner_size(x, y);
+                                         if let Some(wl_surface) = surface.as_ref() {
```
Comment (vberger) [inline src/backend/winit.rs]: To avoid confusion, this is a `wl_egl_surface`, not a `wl_surface`

**Comment by vberger** [comment]
I'd like to have the example running without crashing before merging, as this is likely some weird bug somewhere and we'll probably regret not hunting it if we merge it now.

**Comment by Drakulix** [comment]
Alright, I will hunt it.
No segfault, it just closes and returns a bad error code. Maybe we should integrate `dispatch_new_events` into the `event_loop` somehow. Probably with a timer.

**Comment by vberger** [comment]
I mead, given the source of this example, it finishing by an other way than panicking is very weird, and likely not a good sign at all...

**Comment by Drakulix** [comment]
Well testing on wayland is quite painful rightnow for me, because I need to switch my DE (see https://github.com/Cloudef/wlc/issues/235), which is why I was hoping, that you might have an idea.

**Comment by vberger** [comment]
What I do, is I run weston as an X11 client, and then start my wayland apps as weston clients.

I can try to run your code like this this evening on my computer, if you want.

**Comment by Drakulix** [comment]
> What I do, is I run weston as an X11 client, and then start my wayland apps as weston clients.
Good idea.

> I can try to run your code like this this evening on my computer, if you want.
This will probably be a rather nasty bug, so if you are willing to help debugging (even just by testing), that would be highly appreciated.

File: src/backend/winit.rs [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+
+    let context = unsafe {
+        EGLContext::new(native,
+                        attributes,
+                        PixelFormatRequirements {
+                            hardware_accelerated: Some(true),
+                            color_bits: Some(24),
+                            alpha_bits: Some(8),
+                            ..Default::default()
+                        })?
```
Comment (Drakulix) [inline src/backend/winit.rs]: Okay I have debugged it. This line causes a segfault, if context creation fails (why it fails is a different story, see other comment). The reason is, that the `wl_egl_surface` is dropped, which causes the segfault in this specific case. Any thoughts @vberger ?

File: src/backend/winit.rs [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+/// graphics backend trait, from a given `WindowBuilder` struct and a corresponding
+/// `WinitInputBackend`, which implements the `InputBackend` trait
+pub fn init_from_builder(builder: WindowBuilder)
+                         -> Result<(WinitGraphicsBackend, WinitInputBackend), CreationError> {
+    init_from_builder_with_gl_attr(builder,
+                                   GlAttributes {
+                                       version: None,
+                                       profile: None,
+                                       debug: cfg!(debug_assertions),
+                                       robustness: Robustness::TryRobustLoseContextOnReset,
```
Comment (Drakulix) [inline src/backend/winit.rs]: This causes context creation to fail, because apparently my EGL implementation reports, that it supports robustness, but creation later fails with `BadAttribute`, if it is set to any other value then `NotRobust`.

Comment (vberger) [inline src/backend/winit.rs]: Okay, this is far beyond my knowledge of EGL 😅

File: src/backend/winit.rs (hunk 9 above)
Comment (vberger) [inline src/backend/winit.rs]: if the `wl_egl_surface` is dropped, this destroys the EGL surface, and the pointer you retrieve line 82 becomes invalid. Could this be the reason?

Comment (Drakulix) [inline src/backend/winit.rs]: The crash happens in `wl_egl_window_destroy` though.

File: src/backend/winit.rs (hunk 10 above)
Comment (Drakulix) [inline src/backend/winit.rs]: It just sounds like a typical driver bug. I have changed it to default to `NotRobust` now.

File: src/backend/winit.rs (hunk 9 above)
Comment (vberger) [inline src/backend/winit.rs]: Uh. So, that means the EGL surface ends up in an invalid state following the failed initialization, and then it blows up when it is destroyed, I'd guess?

File: src/backend/graphics/egl.rs [hunk 11]
```diff
@@ -0,0 +1,746 @@
...
+
+        let surface = {
+            let surface = match native {
+                Native::X11(_, window) |
+                Native::Wayland(_, window) |
+                Native::Gbm(_, window) => egl.CreateWindowSurface(display, config_id, window, ptr::null()),
+            };
+
+            if surface.is_null() {
+                return Err(CreationError::OsError(String::from("eglCreateWindowSurface failed")));
```
Comment (vberger) [inline src/backend/graphics/egl.rs]: This check should be done in the `Gbm` branch of the match, otherwise providing a null pointer in input with one of the others backend may trigger it...

An other assert checking that they are not null can be added to replace this one, too.

File: src/backend/winit.rs (hunk 9 above)
Comment (Drakulix) [inline src/backend/winit.rs]: Probably. I will check, what kind of cleanup should be done in this case.

File: src/backend/graphics/egl.rs (hunk 11 above)
Comment (Drakulix) [inline src/backend/graphics/egl.rs]: Are you aware, that the match statement, does the same for all backend types? The rustfmt format might have tricked you there.

Comment (vberger) [inline src/backend/graphics/egl.rs]: Oh! Right, indeed. My bad. 😅

File: src/backend/graphics/egl.rs [hunk 12]
```diff
@@ -0,0 +1,746 @@
...
+            }
+            surface
+        };
```
Comment (vberger) [inline src/backend/graphics/egl.rs]: Good news, I found the cause for the segfault!

The reason is, because of the reentrancy of this function(see the match on line 146, it calls itself), this par ends up called twice. Especially, `egl.CreateWindowSurface()` is called twice on the same native surface, and apparently this is not good.

The simplest fix is to move this whole `let surface = { .... }` block down in the function, right before the end, until after the EGL context has been successfully created (so line 500 of this diff). :)

Comment (vberger) [inline src/backend/graphics/egl.rs]: Note it only solves the segfault, not the panic 😅

**Comment by vberger** [comment]
👍
